solc/CommandLineInterface.cpp

```diff
@@ -1270,6 +1296,26 @@ void CommandLineInterface::assembleYul(yul::YulStack::Language _language, yul::Y
 			else
 				report(Error::Severity::Info, "No text representation found.");
 		}
+		if (m_options.compiler.outputs.asmJson)
+		{
+			std::shared_ptr<evmasm::Assembly> assembly{stack.assembleEVMWithDeployed().first};
+			if (assembly)
+				if (stack.parserResult() && stack.parserResult()->debugData)
+				{
+					std::map<std::string, unsigned> sourceIndices = collectYulSourceIndices(*stack.parserResult());
+					// If sourceIndices are empty, there were no source locations annotated in the yul source.
+					// In this case, we just add the filename of the yul file itself.
+					if (sourceIndices.empty())
+						sourceIndices[src.first] = 0;
+
+					sout() << util::jsonPrint(
+						removeNullMembers(assembly->assemblyJSON(sourceIndices)),
+						m_options.formatting.json
+					) << std::endl;
+					return;
```

Comment:
Good point. I moved the logic to `YulStack`. But I was wondering if placing `collectYulSourceIndices` (now named `sourceIndices`) in `Object.h` might be more appropriate than in `YulStack`. Also, maybe this piece of code: 
```c++
if (sourceIndices.empty())
    sourceIndices[src.first] = 0;
```
Could be moved to `collectYulSourceIndices` as well, I haven't made this adjustment yet, as I am uncertain whether this behavior aligns with the expectations of `sourceIndices` in `CompilerStack`.

Reply:
Maybe. It would make sense as something like `Object::collectSourceIndices()` since it does not really depend on anything in the stack.

But then one might expect it to keep the strings shared via `shared_ptr`. Probably still fine not to, because we can always change it if needed, but the function right now really makes copies like crazy, while the sharing was meant to prevent that so it's a bit against the design of `Object` :) In the current form might be better to just leave it in the stack.